oak-run/src/main/java/org/apache/jackrabbit/oak/run/Downloader.java

```diff
@@ -0,0 +1,263 @@
+                        }
+                    }
+                }).collect(Collectors.toList());
+
+        return new DownloadReport(
+                itemResponses.stream().filter(r -> !r.failed).count(),
+                itemResponses.stream().filter(r -> r.failed).count(),
+                itemResponses.stream().filter(r -> !r.failed).mapToLong(r -> r.size).sum()
+        );
+    }
+
+    @Override
+    public void close() throws IOException {
+        executorService.shutdown();
+    }
+
+    private class DownloaderWorker implements Callable<ItemResponse> {
+
+        private final Item item;
+
+        public DownloaderWorker(Item item) {
+            this.item = item;
+        }
+
+        @Override
+        public ItemResponse call() throws Exception {
+            long t0 = System.nanoTime();
+
+            URLConnection sourceUrl = new URL(item.source).openConnection();
+            sourceUrl.setConnectTimeout(Downloader.this.connectTimeoutMs);
+            sourceUrl.setReadTimeout(Downloader.this.readTimeoutMs);
+
+            Path destinationPath = Paths.get(item.destination);
+            Files.createDirectories(destinationPath.getParent());
+            long size;
+            try (ReadableByteChannel byteChannel = Channels.newChannel(sourceUrl.getInputStream());
+                 FileOutputStream outputStream = new FileOutputStream(destinationPath.toFile())) {
+                size = outputStream.getChannel()
+                        .transferFrom(byteChannel, 0, Long.MAX_VALUE);
+            }
+            long elapsed = TimeUnit.NANOSECONDS.toMillis(System.nanoTime() - t0);
+            if (slowLogThreshold > 0 && elapsed >= slowLogThreshold) {
+                LOG.warn("{} [{}] downloaded in {} ms", item.source, IOUtils.humanReadableByteCount(size), elapsed);
+            } else {
+                LOG.debug("{} [{}] downloaded in {} ms", item.source, IOUtils.humanReadableByteCount(size), elapsed);
+            }
+            return ItemResponse.success(size);
+        }
+
+        @Override
+        public String toString() {
+            return "DownloaderWorker{" +
+                    "item=" + item +
+                    '}';
+        }
+    }
+
+    private static class RetryingCallable<V> implements Callable<V> {
+        private final Callable<V> callable;
+        private final int retries;
+
+        public RetryingCallable(int retries, Callable<V> callable) {
+            this.retries = retries;
+            this.callable = callable;
+        }
+
+        public V call() {
+            int retried = 0;
+            // Save exceptions that are thrown after each failure, so they can be printed if all retries fail
+            List<Throwable> exceptions = new ArrayList<>();
+
+            // Loop until it doesn't throw an exception or max number of tries is reached
+            while (true) {
+                try {
+                    return callable.call();
+                } catch (Exception e) {
```

Comment:
We should have a backoff mechanism. Otherwise we risk doing an excessive number of requests. For instance, if the failure happens at once after a request is sent (so it is not a timeout, like a blob not found or service is down or throttling limits reached), we will be retrying as fast as the network speed allows it. If all requests are failing and we have a high concurrency level, we can easily end up doing hundreds or even thousands of requests per second, which will likely trigger some throttling mechanism on the side of Azure.

I've seen this library being used to implement retries and other fault tolerance mechanisms, maybe it's worth to have a look: https://resilience4j.readme.io/docs Another similar library is Failsafe. There's a nice comparison here: https://blog.frankel.ch/comparison-fault-tolerance-libraries/

Reply:
I have implemented a basic exponential backoff mechanism.